MdeModulePkg/Bus/Isa/Ps2KeyboardDxe/Ps2KbdCtrller.c

```diff
@@ -1807,57 +1807,131 @@ InitKeyboard (
 
 }
 
-
 /**
-  Check whether there is Ps/2 Keyboard device in system by 0xF4 Keyboard Command
-  If Keyboard receives 0xF4, it will respond with 'ACK'. If it doesn't respond, the device
-  should not be in system.
-
-  @param[in]  ConsoleIn             Keyboard Private Data Structure
-
-  @retval     TRUE                  Keyboard in System.
-  @retval     FALSE                 Keyboard not in System.
+  Check if PS2 keyboard is conntected, by sending ECHO command.
+  @param                        none
+  @retval TRUE                  connected
+  @retvar FALSE                 unconnected
 **/
 BOOLEAN
-EFIAPI
-CheckKeyboardConnect (
+DetectPs2Keyboard (
   IN KEYBOARD_CONSOLE_IN_DEV *ConsoleIn
   )
 {
-  EFI_STATUS     Status;
-  UINTN          WaitForValueTimeOutBcakup;
+  UINT32                TimeOut;
+  UINT8                 RegEmptied;
+  UINT8                 Data;
+  UINT32                SumTimeOut;
+  UINT8                 GotIt;
+  UINT8                 Status;
+
+  TimeOut     = 0;
+  RegEmptied  = 0;
 
   //
-  // enable keyboard itself and wait for its ack
-  // If can't receive ack, Keyboard should not be connected.
+  // Wait for input buffer empty
   //
-  if (!PcdGetBool (PcdFastPS2Detection)) {
-    Status = KeyboardWrite (
-               ConsoleIn,
-               KEYBOARD_KBEN
-               );
-
-    if (EFI_ERROR (Status)) {
-      return FALSE;
+  for (TimeOut = 0; TimeOut < KEYBOARD_TIMEOUT; TimeOut += 30) {
+    if ((KeyReadStatusRegister (ConsoleIn) & KEYBOARD_STATUS_REGISTER_HAS_INPUT_DATA) == 0) {
+      RegEmptied = 1;
+      break;
     }
+    MicroSecondDelay (30);
+  }
+
+  if (RegEmptied == 0) {
+    return FALSE;
+  }
+
+  //
+  // Write it
+  //
+  KeyWriteDataRegister (ConsoleIn, KBC_INPBUF_VIA60_KBECHO);
+
+  //
+  // wait for 1s

```

Comment:
outdated comment?

Reply:
It actually waits for 1s, see `SumTimeOut >= KEYBOARD_WAITFORVALUE_TIMEOUT`